download_task.rb

```diff
@@ -0,0 +1,206 @@
+require 'fileutils'
+require 'date'
+
+class DownloadTask
+  attr_reader :provider
+  attr_reader :box_dir
+  attr_reader :agent
+  attr_reader :meta
+
+  def initialize(box_dir, provider)
+    unless File.directory?(box_dir)
+      FileUtils.mkdir_p(box_dir)
+    end
+    @box_dir = box_dir
+    @provider = provider
+
+    @meta_file_name = File.join(box_dir, "#{name}-#{provider}.yml")
+    @meta = read_meta
+
+    @agent = Mechanize.new
+    @agent.follow_meta_refresh = true
+    @agent.ignore_bad_chunking = true
+  end
+
+  def needed?
```

Comment:
This is the base class. There are actually no instances of this. The sub classes override this accordingly. But a safe default is 'true'.